### ISSUE TITLE
handling service registry client errors

### ISSUE THREAD

**Issue by GregHanson** [istio/pilot#1501]
The current interface which the various service discovery clients implement is the following:

```
type ServiceDiscovery interface {
	Services() []*Service
	GetService(hostname string) (*Service, bool)
	Instances(hostname string, ports []string, labels LabelsCollection) []*ServiceInstance
	HostInstances(addrs map[string]bool) []*ServiceInstance
	ManagementPorts(addr string) PortList
}
```

pilot/proxy/envoy/discovery.go calls this interface and updates its cache accordingly.  With the current interface, if the client experiences an error calling the consul/eureka/etc server, it will currently return an empty array, which discovery.go will updates its cache with and return the new empty result to the xDS API's

current suggestions would be to either:

1. add a return error for the interface methods so discovery.go ignores the empty array and returns what it has in the cache

2. caching or better/consistent error handling on the client side model.ServiceDiscovery implementations

Thoughts?

**Comment by rshriram** [comment]
Lets start with 1. And possibly, along the way, split the handlers for consul, eureka, k8s, such that only one of the caches is cleared when service handlers are called, instead of clearing all caches.

There are several low hanging fruits in the cache and we need to have more finer grained cache invalidation.

**Comment by kyessenov** [comment]
The client of the registry already does caching (the discovery service that ultimately constructs the config). So adding another caching to the model adaptor seems redundant. Perhaps, returning an error indicator for "retrying later" is a good option then. I agree that returning an empty array on error is not a good behavior.

**Comment by GregHanson** [comment]
resolved by #1610
